### ISSUE TITLE
With [shared] Of [params]

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/vblang#443]
With [shared] Of [params]

I suggest to use this form to supply Shared methods with all or some of its params:
```vb.net
With IO.Path Of "C:\1.txt"
    Console.WriteLine(.GetDirectory( ))
    Console.WriteLine(.GetFileNameWithoutExtension( ))
    Console.WriteLine(.GetExtension( ))
End With
```

which can be lowered to:
```vb.net
Console.WriteLine(IO.Path.GetDirectory("C:\1.txt"))
Console.WriteLine(P.GetFileNameWithoutExtension("C:\1.txt"))
Console.WriteLine(P.GetExtension("C:\1.txt"))
```

This can be more compact using #442

```vb.net
With IO.Path Of "C:\1.txt",
         Report := Console.WriteLine
    Report(.GetDirectory( ))
    Report(.GetFileNameWithoutExtension( ))
    Report(.GetExtension( ))
End With

**Comment by bandleader** [comment]
This doesn't look feasible. Even if we allow `With` to be used for aliasing classes, which you asked for in #442, here you're trying to do use _two different_ `With`ees and delimiting them with `Of`?

What you probably mean is
```vb
With path = IO.Path, file = "C:\1.txt", Report = Console.WriteLine
  Report(path.GetDirectory(file))
  Report(path.GetFileNameWithoutExtension(file))
  Report(path.GetExtension(file))
End With
```

Using `.GetDirectory` can't work in a multiple-withee `With`, because which one would you be referring to?
Similarly `.GetDirectory( )` can't work -- there's no way for the compiler to guess that you mean to insert the 'withee' in there as an argument.

**Comment by VBAndCs** [comment]
No. There is only on ref (IO.Oath) and on alias (Report). I am building on #438 where With can be used to define a scope for variabls, and als aliases. If there are many vars and aliases, then they should be used, and the dot refers only to the unnamed ref.
The `Of` keyword defines the first param to be sent to the ref method. This is useful to indicate the the block is operating on the same file or data.

**Comment by bandleader** [comment]
1) Do you mean that you want to tell the compiler that inside your `With`, you will have _only_ a list of statements, for which _each_ statement has one innermost method call to a single-parameter function, to which you will not pass any arguments, and you want the compiler to fill in the argument based on the `Of` target of your `With`?
2) If so, can you explain why this is true so often that you would want a compiler to make a special case, rather than just assigning to a name (even `With f =  "C:\1.txt"` and then `path.GetDirectory(f)`?

**Comment by VBAndCs** [comment]
@bandleader 
1. Yes. I was thinking of a way to use some static methods as extenssion methods, like using Path methods as extenssions to the the string c;lass, but there is no easier way to do so. Unless I define an extenssion method and a new non-static Path class to do:
```
With "C:\1.txt".AsPath( )
    Report(.GetDirectory( ))
    Report(.GetFileNameWithoutExtension( ))
    Report(.GetExtension( ))
End With
```

This is OK, but needs to be repeated for every static class that has the similar usage. So, The 'Of param` may be a deneral solution, to use static classes metods as extenssion methods.

2. I can ask you the question about your proposal about using shared methods as extenssions `value.(method)` instead using method(value) :)

**Comment by bandleader** [comment]
>Yes. I was thinking of a way to use some static methods [...]

@VBAndCs Now I understand: what you really want to do is `With New IO.Path("C:\1.txt")` and then `.GetDirectory()` etc. But that's not the way IO.Path works; it's a collection of path methods, not an instance class. If you want an instance class for this, you are welcome to create `VBAndCs.IO.Path` that works the way you describe. But static classes are meant to be static classes. This is not a candidate for a language feature IMHO. Remember also that not every method there has to have the same signature `(path As String)`. 

>I can ask you the question about your proposal about using shared methods as extenssions value.(method) instead using method(value) :)

@VBAndCs The motivations for piping are clearly explained in its issue #165, which you are welcome to read. One the primary ones is the need to preserve fluency/execution order/reduce back tracking, described more generally in #154, and actually shared with extension methods, fluent casting, and other features. These were ideas spearheaded by outgoing Visual Basic Program Manager Anthony Green.

**Comment by VBAndCs** [comment]
This example may be clearer:
```
With IO.Path Of "C:\Test"
     Console.WriteLine(.Combine("1.txt"))
End With
```

This is equivalent to:
```
Console.WriteLine(IO.Path.Combine("C:\Test", "1.txt"))
```

Another one:
```
With IO.Path Of ("C:\Test", "1.txt")
     Console.WriteLine(.Combine())
End With
```

The idea is to create a param list that can be reused many times in the block if possible. Of course the chance to reuse one param for ,any times is bigger than reusing two params.

**Comment by bandleader** [comment]
Please read the above. Thanks.

**Comment by VBAndCs** [comment]
I already answered before:
> This is OK, but needs to be repeated for every static class that has the similar usage. So, The 'Of param` may be a general solution, to use static classes metods as extenssion methods.
